### PR TITLE
feat(regressor): add regressor interface

### LINKED ISSUE #14: Feature Requests
This issue collects all features requests. Any one is welcomed to work on issues listed below, and do not forget to include your contributions and name in the `CHANGELOG.rst`.

If you want to work on a requested feature, please re-open the linked issue, and leave a comment below to let us know that you want to work on it.

### New features
* [x] `CascadeForestRegressor` class for regression problem (#4)
* [ ] `export_graphviz` method on visualizing decision trees in deep forest (#12)
* [x] Internal label encoder on the class labels (#13)
* [x] Better support on the input data (#19, #20)
* [ ] GPU Support (#27)
* [ ] Support customized estimator in the layers (#29)

### Python package
* [ ] Build wheels for Python 2.7

### New language wrappers:
* [ ] C/C++ Interface (#9)

### PR DESCRIPTION
**Pull request by tczhao**
add regressor interface

resolves #14 #4

### CODE REVIEW

**Comment by xuyxu** [comment]
Brilliant job! Thanks @tczhao! I will take a careful look tomorrow ;-)

**Comment by dswool** [comment]
The classification performance on our private data is soso but the regression is incredible

**Comment by tczhao** [comment]
The classification performance on our private data is soso but the regression is incredible

File: deepforest/cascade.py [hunk 1]
```diff
@@ -298,8 +492,10 @@ def __getitem__(self, index):
 
     def _get_n_output(self, y):
         """Return the number of output inferred from the training labels."""
-        n_output = np.unique(y).shape[0]  # classification
-        return n_output
+        if is_classifier(self):
+            n_output = np.unique(y).shape[0]  # classification
+            return n_output
+        return y.shape[0]
```
Comment (xuyxu) [inline deepforest/cascade.py:498]: Setting `self.n_outputs_` to `y.shape[0]` is problematic, for example, on the boston dataset, `y_train` is of the shape `(379,)`. How about add another IF statement to distinguish univariate regression from multi-variate one? Something like:

```
if len(y.shape) == 0:
    return 1  # univariate regression
else:
    return y.shape[1]  # multi-variate regression
```

File: docs/index.rst [hunk 2]
```diff
@@ -53,6 +53,23 @@ Quickstart
...
+    from sklearn.metrics import mean_squared_error
+
+    from deepforest import CascadeForestRegressor
+
+    X, y = load_boston(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
+    model = CascadeForestRegressor(random_state=1)
+    model.fit(X_train, y_train)
+    y_pred = model.predict(X_test)
+    acc = mean_squared_error(y_test, y_pred)
```
Comment (xuyxu) [inline docs/index.rst:69]: changing the variable name from `acc` to `mse` would be better ;-)

File: deepforest/cascade.py [hunk 3]
```diff
@@ -1062,3 +1290,139 @@ def predict(self, X):
...
+
+    @deepforest_model_doc(
+        """Build a deep forest using the training data.""", "regressor_fit"
+    )
+    def fit(self, X, y, sample_weight=None):
+        super().fit(X, y, sample_weight)
+
+    def predict(self, X):
+        """
+        Predict class for X.
```
Comment (xuyxu) [inline deepforest/cascade.py:1351]: The docstrings for regressor predict should be `Predict regression target for X.`

**Review by xuyxu — COMMENTED** [review]
Hi @tczhao, I have left some comments on your PR, let me know if you are OK with them. Thanks for this great PR, I will work on the experiment parts on the regressor and let you stay stuned. Hopefully, this PR can be merged in a few days.

**Comment by xuyxu** [comment]
> The classification performance on our highly skewed data is not that good but the regression performance is incredible

Thanks for sharing this. Another question is that what is the final number of cascade layers in your regression dataset? If the number of 1, I am afraid that the cascade structure does not have any positive effect, and the entire model is simply an average of forest models...

**Comment by tczhao** [comment]
> > The classification performance on our highly skewed data is not that good but the regression performance is incredible
> 
> Thanks for sharing this. Another question is that what is the final number of cascade layers in your regression dataset? If the number of 1, I am afraid that the cascade structure does not have any positive effect, and the entire model is simply an average of forest models...
> 
> **EDIT**: I have added the experiment on regression, you can check it via the last GitHub checks

Thanks for the added tests

It's 3 layer in the end (not just a simple ensemble), we set the max to 3 for some quick experiment

File: deepforest/cascade.py (hunk 1 above)
Comment (tczhao) [reply deepforest/cascade.py]: turned out this output is not getting used in downstream function regressor, set default to 1 and added a comment

for multiple output regressor, maybe we can leave this as a next step

There's some nice thread here on have using a MultiOutputRegressor wrapper
https://github.com/dmlc/xgboost/issues/2087

File: docs/index.rst (hunk 2 above)
Comment (tczhao) [reply docs/index.rst]: Realise the print statement variable also needs to change, will fix this tomorrow

**Review by xuyxu — COMMENTED** [review]
LGTM

**Comment by xuyxu** [comment]
Thanks for your great work @tczhao! I will merge the PR soon ;-)

**Comment by xuyxu** [comment]
Merged.
